File: packages/client-py/src/durable_streams/_sse.py
"""
Server-Sent Events (SSE) parsing for the Durable Streams protocol.

This module handles parsing of SSE event streams according to the protocol:
- `event: data` events contain the stream data
- `event: control` events contain `streamNextOffset` and optional `streamCursor`
"""

import codecs
import json
from collections.abc import AsyncIterator, Iterator
from dataclasses import dataclass

from durable_streams._errors import DurableStreamError
from durable_streams._types import Offset


@dataclass
class SSEDataEvent:
    """An SSE data event containing stream data."""

    type: str = "data"
    data: str = ""


@dataclass
class SSEControlEvent:
    """
    An SSE control event with stream metadata.

    Attributes:
        stream_next_offset: The next offset to read from
        stream_cursor: Optional cursor for CDN collapsing
        up_to_date: Whether caught up to end of stream
        stream_closed: Whether the stream is closed (EOF)
    """

    type: str = "control"
    stream_next_offset: Offset = ""
    stream_cursor: str | None = None
    up_to_date: bool = False
    stream_closed: bool = False


SSEEvent = SSEDataEvent | SSEControlEvent


class SSEParser:
    """
    Incremental SSE parser.

    Maintains state for parsing SSE events from a byte stream.
    """

    def __init__(self) -> None:
        self._buffer = ""
        self._current_event_type: str | None = None
        self._current_data: list[str] = []

    def feed(self, chunk: str) -> list[SSEEvent]:
        """
        Feed a chunk of data and return any complete events.

        Args:
            chunk: String chunk to parse

        Returns:
            List of complete SSE events
        """
        self._buffer += chunk
        events: list[SSEEvent] = []

        # Process complete lines
        while "\n" in self._buffer:
            line, self._buffer = self._buffer.split("\n", 1)

            # Empty line signals end of event
            if line == "" or line == "\r":
                event = self._emit_event()
                if event is not None:
                    events.append(event)
                continue

            # Strip trailing CR if present
            if line.endswith("\r"):
                line = line[:-1]

            # Parse the line
            if line.startswith("event:"):
                # Per SSE spec, strip only one optional space after "event:"
                event_type = line[6:]
                if event_type.startswith(" "):
                    event_type = event_type[1:]
                self._current_event_type = event_type
            elif line.startswith("data:"):
                # Per SSE spec, strip the optional space after "data:"
                content = line[5:]
                if content.startswith(" "):
                    content = content[1:]
                self._current_data.append(content)
            # Ignore other fields (id, retry, comments starting with :)

        return events

    def _emit_event(self) -> SSEEvent | None:
        """Emit the current event if valid."""
        if self._current_event_type is None or not self._current_data:
            self._reset()
            return None

        data_str = "\n".join(self._current_data)
        event_type = self._current_event_type
        self._reset()

        if event_type == "data":
            return SSEDataEvent(data=data_str)

        if event_type == "control":
            try:
                control = json.loads(data_str)
                stream_closed = control.get("streamClosed", False)
                up_to_date = control.get("upToDate", False) or stream_closed
                return SSEControlEvent(
                    stream_next_offset=control.get("streamNextOffset", ""),
                    stream_cursor=control.get("streamCursor"),
                    up_to_date=up_to_date,
                    stream_closed=stream_closed,
                )
            except json.JSONDecodeError as e:
                # Control events contain critical offset data - don't silently ignore
                preview = data_str[:100] + "..." if len(data_str) > 100 else data_str
                raise DurableStreamError(
                    f"Failed to parse SSE control event: {e}. Data: {preview}",
                    code="PARSE_ERROR",
                ) from e

        return None

    def _reset(self) -> None:
        """Reset parser state for next event."""
        self._current_event_type = None
        self._current_data = []

    def finish(self) -> list[SSEEvent]:
        """
        Finish parsing and return any remaining events.

        Call this when the stream ends to handle any buffered data.

        Returns:
            List of any remaining complete events
        """
        events: list[SSEEvent] = []

        # Process remaining buffer - add double newline to flush event
        if self._buffer:
            remaining_events = self.feed("\n\n")
            events.extend(remaining_events)
            self._buffer = ""
        elif self._current_event_type is not None and self._current_data:
            # We have a partial event waiting - emit it
            event = self._emit_event()
            if event is not None:
                events.append(event)

        return events


def parse_sse_sync(byte_iterator: Iterator[bytes]) -> Iterator[SSEEvent]:
    """
    Parse SSE events from a synchronous byte iterator.

    Uses incremental UTF-8 decoding to handle multi-byte characters
    that may be split across chunk boundaries.

    Args:
        byte_iterator: Iterator yielding bytes

    Yields:
        Parsed SSE events
    """
    parser = SSEParser()
    # Use incremental decoder to handle UTF-8 codepoints split across chunks
    decoder = codecs.getincrementaldecoder("utf-8")("replace")

    for chunk in byte_iterator:
        text = decoder.decode(chunk)
        if text:
            for event in parser.feed(text):
                yield event

    # Flush any remaining bytes in the decoder
    final_text = decoder.decode(b"", final=True)
    if final_text:
        for event in parser.feed(final_text):
            yield event

    # Handle any remaining events in the parser
    for event in parser.finish():
        yield event


async def parse_sse_async(
    byte_iterator: AsyncIterator[bytes],
) -> AsyncIterator[SSEEvent]:
    """
    Parse SSE events from an asynchronous byte iterator.

    Uses incremental UTF-8 decoding to handle multi-byte characters
    that may be split across chunk boundaries.

    Args:
        byte_iterator: Async iterator yielding bytes

    Yields:
        Parsed SSE events
    """
    parser = SSEParser()
    # Use incremental decoder to handle UTF-8 codepoints split across chunks
    decoder = codecs.getincrementaldecoder("utf-8")("replace")

    async for chunk in byte_iterator:
        text = decoder.decode(chunk)
        if text:
            for event in parser.feed(text):
                yield event

    # Flush any remaining bytes in the decoder
    final_text = decoder.decode(b"", final=True)
    if final_text:
        for event in parser.feed(final_text):
            yield event

    # Handle any remaining events in the parser
    for event in parser.finish():
        yield event

File: packages/client/src/sse.ts
/**
 * SSE (Server-Sent Events) parsing utilities for the durable streams protocol.
 *
 * SSE format from protocol:
 * - `event: data` events contain the stream data
 * - `event: control` events contain `streamNextOffset` and optional `streamCursor` and `upToDate`
 */

import { DurableStreamError } from "./error"
import type { Offset } from "./types"

/**
 * Parsed SSE event from the stream.
 */
export interface SSEDataEvent {
  type: `data`
  data: string
}

export interface SSEControlEvent {
  type: `control`
  streamNextOffset: Offset
  streamCursor?: string
  upToDate?: boolean
  streamClosed?: boolean
}

export type SSEEvent = SSEDataEvent | SSEControlEvent

/**
 * Parse SSE events from a ReadableStream<Uint8Array>.
 * Yields parsed events as they arrive.
 */
export async function* parseSSEStream(
  stream: ReadableStream<Uint8Array>,
  signal?: AbortSignal
): AsyncGenerator<SSEEvent, void, undefined> {
  const reader = stream.getReader()
  const decoder = new TextDecoder()
  let buffer = ``
  let currentEvent: { type?: string; data: Array<string> } = { data: [] }

  try {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    while (true) {
      if (signal?.aborted) {
        break
      }

      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })

      // Normalize line endings: CRLF → LF, lone CR → LF (per SSE spec)
      buffer = buffer.replace(/\r\n/g, `\n`).replace(/\r/g, `\n`)

      // Process complete lines
      const lines = buffer.split(`\n`)
      // Keep the last incomplete line in the buffer
      buffer = lines.pop() ?? ``

      for (const line of lines) {
        if (line === ``) {
          // Empty line signals end of event
          if (currentEvent.type && currentEvent.data.length > 0) {
            const dataStr = currentEvent.data.join(`\n`)

            if (currentEvent.type === `data`) {
              yield { type: `data`, data: dataStr }
            } else if (currentEvent.type === `control`) {
              try {
                const control = JSON.parse(dataStr) as {
                  streamNextOffset: Offset
                  streamCursor?: string
                  upToDate?: boolean
                  streamClosed?: boolean
                }
                yield {
                  type: `control`,
                  streamNextOffset: control.streamNextOffset,
                  streamCursor: control.streamCursor,
                  upToDate: control.upToDate,
                  streamClosed: control.streamClosed,
                }
              } catch (err) {
                // Control events contain critical offset data - don't silently ignore
                const preview =
                  dataStr.length > 100 ? dataStr.slice(0, 100) + `...` : dataStr
                throw new DurableStreamError(
                  `Failed to parse SSE control event: ${err instanceof Error ? err.message : String(err)}. Data: ${preview}`,
                  `PARSE_ERROR`
                )
              }
            }
            // Unknown event types are silently skipped per protocol
          }
          currentEvent = { data: [] }
        } else if (line.startsWith(`event:`)) {
          // Per SSE spec, strip only one optional space after "event:"
          const eventType = line.slice(6)
          currentEvent.type = eventType.startsWith(` `)
            ? eventType.slice(1)
            : eventType
        } else if (line.startsWith(`data:`)) {
          // Per SSE spec, strip the optional space after "data:"
          const content = line.slice(5)
          currentEvent.data.push(
            content.startsWith(` `) ? content.slice(1) : content
          )
        }
        // Ignore other fields (id, retry, comments)
      }
    }

    // Handle any remaining data
    const remaining = decoder.decode()
    if (remaining) {
      buffer += remaining
    }

    // Process any final event
    if (buffer && currentEvent.type && currentEvent.data.length > 0) {
      const dataStr = currentEvent.data.join(`\n`)
      if (currentEvent.type === `data`) {
        yield { type: `data`, data: dataStr }
      } else if (currentEvent.type === `control`) {
        try {
          const control = JSON.parse(dataStr) as {
            streamNextOffset: Offset
            streamCursor?: string
            upToDate?: boolean
            streamClosed?: boolean
          }
          yield {
            type: `control`,
            streamNextOffset: control.streamNextOffset,
            streamCursor: control.streamCursor,
            upToDate: control.upToDate,
            streamClosed: control.streamClosed,
          }
        } catch (err) {
          const preview =
            dataStr.length > 100 ? dataStr.slice(0, 100) + `...` : dataStr
          throw new DurableStreamError(
            `Failed to parse SSE control event: ${err instanceof Error ? err.message : String(err)}. Data: ${preview}`,
            `PARSE_ERROR`
          )
        }
      }
    }
  } finally {
    reader.releaseLock()
  }
}
